xls/solvers/z3_ir_equivalence.cc

```diff
@@ -46,6 +49,92 @@ namespace xls::solvers::z3 {
 
 namespace {
 
+std::optional<std::string> GetAssertKey(const Assert* asrt) {
+  if (asrt->label().has_value()) {
+    return asrt->label();
+  }
+  if (asrt->original_label().has_value()) {
+    return asrt->original_label();
+  }
+  return std::nullopt;
+}
+
+absl::StatusOr<std::vector<Node*>> BuildAssertChecks(
+    Function* f, const absl::flat_hash_map<Node*, Node*>& cloned_from_b) {
+  absl::flat_hash_map<std::string, Assert*> original_asserts;
+  absl::flat_hash_map<std::string, Assert*> transformed_asserts;
+  absl::flat_hash_set<Node*> b_nodes;
+  b_nodes.reserve(cloned_from_b.size());
+  for (const auto& [_, clone] : cloned_from_b) {
+    b_nodes.insert(clone);
+  }
+
+  for (Node* node : f->nodes()) {
+    if (!node->Is<Assert>()) {
+      continue;
+    }
+    if (b_nodes.contains(node)) {
+      continue;
+    }
+    std::optional<std::string> key = GetAssertKey(node->As<Assert>());
+    if (!key.has_value()) {
+      continue;
+    }
+    if (!original_asserts.emplace(*key, node->As<Assert>()).second) {
+      return absl::InvalidArgumentError(absl::StrFormat(
+          "Duplicate assert label `%s` in original function", *key));
+    }
+  }
+
+  for (const auto& [original_node, clone_node] : cloned_from_b) {
+    if (!original_node->Is<Assert>()) {
+      continue;
+    }
+    std::optional<std::string> key =
+        GetAssertKey(original_node->As<Assert>());
+    if (!key.has_value()) {
+      continue;
+    }
+    if (!transformed_asserts.emplace(*key, clone_node->As<Assert>()).second) {
+      return absl::InvalidArgumentError(absl::StrFormat(
+          "Duplicate assert label `%s` in transformed function", *key));
+    }
+  }
+
+  std::vector<Node*> checks;
+  checks.reserve(original_asserts.size() + transformed_asserts.size());
+  for (const auto& [label, original_assert] : original_asserts) {
+    auto transformed_it = transformed_asserts.find(label);
+    if (transformed_it != transformed_asserts.end()) {
+      Node* cond_a =
+          original_assert->operand(Assert::kConditionOperand);
+      Node* cond_b =
+          transformed_it->second->operand(Assert::kConditionOperand);
+      checks.push_back(f->AddNode(std::make_unique<CompareOp>(
```

Comment:
We usually add new nodes like

```
XLS_ASSIGN_OR_RETURN(std::back_inserter(checks), f->MakeNode<CompareOp>(SourceInfo, cond_a, cond_b, Op::Eq));
```

or

```
XLS_ASSIGN_OR_RETURN(std::back_inserter(checks), f->MakeNodeWithName<CompareOp>(SourceInfo, cond_a, cond_b, Op::Eq, absl::StrCat("assert_", label, "_match")));
```

here and below.

The MakeNode path will also invoke the verifier to do some basic structural checks of the node being added which can catch some bugs.